rfcs/2021-02-28-streams.md

```diff
@@ -0,0 +1,47 @@
+# Streams
+
+Managing dedicated SMP queues for fast synchronous communication
+
+## Problem
+
+SMP agent protocol implementation provides the most secure way to distribute keys achieving the following qualities:
+
+- compromising the sender agent/device allows to send messages, but not to read them
+- compromising the recipient agent/device allows to receive messages, but not to send them (neither public encryption key is stored - TBC if public key can be restored from the private - nor server authentication key is available)
+- compromising the server does not expose any information about messages content, as encrypted section has the same size
+
+The current hybrid encryption scheme uses a new symmetric key for each message, and because the symmetric key is not persisted at any point it is difficult to obtain it and send counterfeit messages on behalf of the sender (in case both the server and recipient are compromised).
+
+There are 2 downsides of the current scheme:
+
+1. RSA encryption/decryption is relatively slow (we use large public exponents) even for 2048 key size and gets slower for larger key sizes. It is is not a problem for the chat messages (and any content updates) that happen infrequently (even if it is every second) but it would slow the transmission of large files and any other streaming communication.
+
+2. If the large file or voice/video calls were to be sent via the same queue as normal messages/content updates, the server would be able to understand by observing the traffic when such transmissions happen (even if performance was not a problem).
+
+## Proposed solution
+
+For every fast/dense data transmission use a separate short-lived unidirectional SMP queue that the receiving agent would provision to receive such transmission and communicate queue address (server and queue ID).
+
+Such queues can be managed by agents via additional set of agent commands, without using connection abstraction, but instead using "stream" abstraction.
+
+The protocol to create and manage such stream would be the following:
+
+1. The sender client notifies the agent about the stream it has to create for a specific duplex connection (the connection must be duplex for it to be possible): `<conn_alias> SNEW <name> <size>` (where `conn_alias` is the existing duplex connection, `size` is the optional size for the data to be transferred; 0 for unbounded streams).
+2. The sender agent sends to the recipient agent the "envelope" `SNEW <name> <size>`
+3. The receiving agent would notify the client about availability of the stream by sending `<conn_alias> STREAM <name> <size>` notification.
```

Comment:
for consistency with later steps - should `name` be renamed to `stream_id`? or were these implied to be different entities?

Reply:
It is not well thought through. The idea was that the name is descriptive, but possibly it is not needed, as probably some additional information about the stream needs to be added by the clients as a normal message anyway, so that the client can decide to accept or not. In this case SNEW would not need to have the name at all, instead the agent would respond with stream ID. Same problem as with current connection name/ID.

I think we should postpone finalising it until chat protocol formalization.

Reply:
removed names